 <b> Team Members</b>
 
 <i> Sandip Pandey</i>
 
 <i> Shree Sapkota</i>
 
 <i> Thuong Nguyen</i>

 # Dataset (Statistics Finland)
 Statistics Finland combines collected data with its own expertise to produce statistics and information services.
Founded in 1865, Statistics Finland is the only Finnish public authority specifically established for statistics. It produces the vast majority of Finnish official statistics and is a significant international actor in the field of statistics.

We collected the Immigrants and integration dataset from there vast database collections for data analysis purposes. This dataset collects information on foreigners visiting Finland throughout the years depending on the regions of the country as well as overall whole country. For data analysis purposes, we have gathered information on Citizens of foreign countries, Persons born abroad, Foreign-language speakers and Persons with foreign background depending on the regions.

## Regions of Finland
- WHOLE COUNTRY
- Uusimaa
- Southwest
- Satakunta
- Kanta-Häme
- Pirkanmaa
- Päijät-Häme
- Kymenlaakso
- South Karelia
- South Savo
- North Savo
- North Karelia
- Central
- South Ostrobothnia
- Ostrobothnia
- Central Ostrobothnia
- North Ostrobothnia
- Kainuu
- Lapland
- Åland

## <span style="color:green">Variables of the dataset based on regions of Finland</span>
- Citizens of foreign countries

- Persons born abroad

- Foreign-language speakers

- Persons with foreign background
- Share of foreign citizens, %

- Share of persons born abroad, %
- Share of foreign-language speakers, %
- Share of persons with foreign background , %
- Share of foreign citizens among foreign citizens in the whole country, %
- Share of persons born abroad among persons born abroad in the whole country, %
- Share of foreign-language speakers among foreign-language speakers in the whole country, %
- The foreign languages are other than Finnish, Swedish and Sami
- Share of persons with foreign background among persons with foreign backgr. in the whole country, %


# Analysis
<b>Untidy part</b>
- Splitting the dataset on basis of Regions to see the different values.
- Checking the maximum values for each regions foreigners information.
- Plotting the graphs to see the trend of the foreigners moving to Finland.
- Plotting the plotly graph according to the dataset.

<b>Tidy part</b>
- Grouping the dataset according to the Regions and foreigners informations.
- Checking different years data according to the Regions.
- Visualization of the tidy data to see different trends of the foreigners coming to Finland.




In [798]:
#importing all the libraries
import pandas as pd
import numpy as np
#importing all the necessary libraries for visualization
import matplotlib.pyplot as plt
import ipywidgets as widgets
from ipywidgets import interact
import plotly.graph_objects as go
import cufflinks as cf
from plotly.offline import iplot,init_notebook_mode

init_notebook_mode(connected=True)
cf.go_offline(connected=True)


#  Loading the csv file

In [799]:
# Loading the csv file
data=pd.read_csv('PopFinland.csv',sep=';',header=[0],skiprows=[0],encoding='latin-1')

In [800]:
data

Year  WHOLE COUNTRY Citizens of foreign countries  \
0  2010                                       167954   
1  2011                                       183133   
2  2012                                       195511   
3  2013                                       207511   
4  2014                                       219675   
5  2015                                       229765   
6  2016                                       243639   
7  2017                                       249452   
8  2018                                       257572   
9  2019                                       267629   

   WHOLE COUNTRY Persons born abroad  WHOLE COUNTRY Foreign-language speakers  \
0                             248135                                   224388   
1                             266148                                   244827   
2                             285471                                   266949   
3                             304279                                   289068   
4                             321977                                   310306   
5                             337162                                   329562   
6                             357541                                   353993   
7                             372802                                   373325   
8                             387215                                   391746   
9                             404179                                   412644   

   WHOLE COUNTRY Persons with foreign background  \
0                                         237066   
1                                         257494   
2                                         279616   
3                                         301524   
4                                         322711   
5                                         339925   
6                                         364787   
7                                         384123   
8                                         402619   
9                                         423494   

   WHOLE COUNTRY Share of foreign citizens, %  \
0                                         3.1   
1                                         3.4   
2                                         3.6   
3                                         3.8   
4                                         4.0   
5                                         4.2   
6                                         4.4   
7                                         4.5   
8                                         4.7   
9                                         4.8   

   WHOLE COUNTRY Share of persons born abroad, %  \
0                                            4.6   
1                                            4.9   
2                                            5.3   
3                                            5.6   
4                                            5.9   
5                                            6.1   
6                                            6.5   
7                                            6.8   
8                                            7.0   
9                                            7.3   

   WHOLE COUNTRY Share of foreign-language speakers, %  \
0                                                4.2     
1                                                4.5     
2                                                4.9     
3                                                5.3     
4                                                5.7     
5                                                6.0     
6                                                6.4     
7                                                6.8     
8                                                7.1     
9                                                7.5     

   WHOLE COUNTRY Share of persons with foreign background , %  \
0                                                4.4            
1                                      

In [801]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Columns: 265 entries, Year to MK21 Åland Share of persons with foreign background among persons with foreign backgr. in the whole country, %
dtypes: float64(176), int64(89)
memory usage: 20.8 KB


<b>This dataset includes too many columns whcih will make the data analysis confusing. Therefore, we decided to stick with only fewer columns so we can analyize more locations data. It was decided that all the percentage value to be dropped from the dataset so we will only have data representing the whole numbers. Here, we have removed the unnecessary columns using % sign as a pattern to remove all the columns which have % sign since these columns are not necessary for our data analysis purposes.</b>

In [802]:
data=data[data.columns[~data.columns.str.contains(r'\%$')]]
data

Year  WHOLE COUNTRY Citizens of foreign countries  \
0  2010                                       167954   
1  2011                                       183133   
2  2012                                       195511   
3  2013                                       207511   
4  2014                                       219675   
5  2015                                       229765   
6  2016                                       243639   
7  2017                                       249452   
8  2018                                       257572   
9  2019                                       267629   

   WHOLE COUNTRY Persons born abroad  WHOLE COUNTRY Foreign-language speakers  \
0                             248135                                   224388   
1                             266148                                   244827   
2                             285471                                   266949   
3                             304279                                   289068   
4                             321977                                   310306   
5                             337162                                   329562   
6                             357541                                   353993   
7                             372802                                   373325   
8                             387215                                   391746   
9                             404179                                   412644   

   WHOLE COUNTRY Persons with foreign background  \
0                                         237066   
1                                         257494   
2                                         279616   
3                                         301524   
4                                         322711   
5                                         339925   
6                                         364787   
7                                         384123   
8                                         402619   
9                                         423494   

   MA1 MAINLAND FINLAND Citizens of foreign countries  \
0                                                  0    
1                                                  0    
2                                                  0    
3                                                  0    
4                                                  0    
5                                                  0    
6                                                  0    
7                                                  0    
8                                                  0    
9                                                  0    

   MA1 MAINLAND FINLAND Persons born abroad  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   
5                                         0   
6                                         0   
7                                         0   
8                                         0   
9                                         0   

   MA1 MAINLAND FINLAND Foreign-language speakers  \
0                                               0   
1                                               0   
2                                               0   
3                                               0   
4                                               0   
5                                               0   
6                                               0   
7                                               0   
8                                               0   
9                                               0   

   MA1 MAINLAND FINLAND Persons with foreign background  \
0                                                  0      
1                                                  0      
2             

 Now, the data set is reduced to 89 columns from 265 columns.

In [803]:
data.columns
data.shape

(10, 89)

In [804]:
data=data.T
data

0       1       2       3  \
Year                                             2010    2011    2012    2013   
WHOLE COUNTRY Citizens of foreign countries    167954  183133  195511  207511   
WHOLE COUNTRY Persons born abroad              248135  266148  285471  304279   
WHOLE COUNTRY Foreign-language speakers        224388  244827  266949  289068   
WHOLE COUNTRY Persons with foreign background  237066  257494  279616  301524   
...                                               ...     ...     ...     ...   
MA2 ÅLAND Persons with foreign background           0       0       0       0   
MK21 Åland Citizens of foreign countries         2415    2560    2610    2721   
MK21 Åland Persons born abroad                   3785    4018    4171    4316   
MK21 Åland Foreign-language speakers             1462    1623    1733    1850   
MK21 Åland Persons with foreign background       3054    3293    3454    3618   

                                                    4       5       6       7  \
Year                                             2014    2015    2016    2017   
WHOLE COUNTRY Citizens of foreign countries    219675  229765  243639  249452   
WHOLE COUNTRY Persons born abroad              321977  337162  357541  372802   
WHOLE COUNTRY Foreign-language speakers        310306  329562  353993  373325   
WHOLE COUNTRY Persons with foreign background  322711  339925  364787  384123   
...                                               ...     ...     ...     ...   
MA2 ÅLAND Persons with foreign background           0       0       0       0   
MK21 Åland Citizens of foreign countries         2837    2927    3092    3182   
MK21 Åland Persons born abroad                   4502    4642    4905    5082   
MK21 Åland Foreign-language speakers             1986    2093    2255    2388   
MK21 Åland Persons with foreign background       3824    3862    4126    4316   

                                                    8       9  
Year                                             2018    2019  
WHOLE COUNTRY Citizens of foreign countries    257572  267629  
WHOLE COUNTRY Persons born abroad              387215  404179  
WHOLE COUNTRY Foreign-language speakers        391746  412644  
WHOLE COUNTRY Persons with foreign background  402619  423494  
...                                               ...     ...  
MA2 ÅLAND Persons with foreign background           0       0  
MK21 Åland Citizens of foreign countries         3396    3421  
MK21 Åland Persons born abroad                   5377    5497  
MK21 Åland Foreign-language speakers             2538    2621  
MK21 Åland Persons with foreign background       4612    4758  

[89 rows x 10 columns]

 Data is transposed to change the rows and columns value for further analysis purposes.

<b> Resetting the index </b>

In [805]:
#resetting the index
data.reset_index(inplace=True)
data

index       0       1       2  \
0                                            Year    2010    2011    2012   
1     WHOLE COUNTRY Citizens of foreign countries  167954  183133  195511   
2               WHOLE COUNTRY Persons born abroad  248135  266148  285471   
3         WHOLE COUNTRY Foreign-language speakers  224388  244827  266949   
4   WHOLE COUNTRY Persons with foreign background  237066  257494  279616   
..                                            ...     ...     ...     ...   
84      MA2 ÅLAND Persons with foreign background       0       0       0   
85       MK21 Åland Citizens of foreign countries    2415    2560    2610   
86                 MK21 Åland Persons born abroad    3785    4018    4171   
87           MK21 Åland Foreign-language speakers    1462    1623    1733   
88     MK21 Åland Persons with foreign background    3054    3293    3454   

         3       4       5       6       7       8       9  
0     2013    2014    2015    2016    2017    2018    2019  
1   207511  219675  229765  243639  249452  257572  267629  
2   304279  321977  337162  357541  372802  387215  404179  
3   289068  310306  329562  353993  373325  391746  412644  
4   301524  322711  339925  364787  384123  402619  423494  
..     ...     ...     ...     ...     ...     ...     ...  
84       0       0       0       0       0       0       0  
85    2721    2837    2927    3092    3182    3396    3421  
86    4316    4502    4642    4905    5082    5377    5497  
87    1850    1986    2093    2255    2388    2538    2621  
88    3618    3824    3862    4126    4316    4612    4758  

[89 rows x 11 columns]

In [806]:
data.columns=data.loc[0]
data

0                                            Year    2010    2011    2012  \
0                                            Year    2010    2011    2012   
1     WHOLE COUNTRY Citizens of foreign countries  167954  183133  195511   
2               WHOLE COUNTRY Persons born abroad  248135  266148  285471   
3         WHOLE COUNTRY Foreign-language speakers  224388  244827  266949   
4   WHOLE COUNTRY Persons with foreign background  237066  257494  279616   
..                                            ...     ...     ...     ...   
84      MA2 ÅLAND Persons with foreign background       0       0       0   
85       MK21 Åland Citizens of foreign countries    2415    2560    2610   
86                 MK21 Åland Persons born abroad    3785    4018    4171   
87           MK21 Åland Foreign-language speakers    1462    1623    1733   
88     MK21 Åland Persons with foreign background    3054    3293    3454   

0     2013    2014    2015    2016    2017    2018    2019  
0     2013    2014    2015    2016    2017    2018    2019  
1   207511  219675  229765  243639  249452  257572  267629  
2   304279  321977  337162  357541  372802  387215  404179  
3   289068  310306  329562  353993  373325  391746  412644  
4   301524  322711  339925  364787  384123  402619  423494  
..     ...     ...     ...     ...     ...     ...     ...  
84       0       0       0       0       0       0       0  
85    2721    2837    2927    3092    3182    3396    3421  
86    4316    4502    4642    4905    5082    5377    5497  
87    1850    1986    2093    2255    2388    2538    2621  
88    3618    3824    3862    4126    4316    4612    4758  

[89 rows x 11 columns]

<b>Here, we can see that there are 2 year rows.So, we have dropped the extra year row from the dataset below using drop().</b>

In [807]:
data=data.drop(0)
data

0                                                Year    2010    2011    2012  \
1         WHOLE COUNTRY Citizens of foreign countries  167954  183133  195511   
2                   WHOLE COUNTRY Persons born abroad  248135  266148  285471   
3             WHOLE COUNTRY Foreign-language speakers  224388  244827  266949   
4       WHOLE COUNTRY Persons with foreign background  237066  257494  279616   
5   MA1 MAINLAND FINLAND Citizens of foreign count...       0       0       0   
..                                                ...     ...     ...     ...   
84          MA2 ÅLAND Persons with foreign background       0       0       0   
85           MK21 Åland Citizens of foreign countries    2415    2560    2610   
86                     MK21 Åland Persons born abroad    3785    4018    4171   
87               MK21 Åland Foreign-language speakers    1462    1623    1733   
88         MK21 Åland Persons with foreign background    3054    3293    3454   

0     2013    2014    2015    2016    2017    2018    2019  
1   207511  219675  229765  243639  249452  257572  267629  
2   304279  321977  337162  357541  372802  387215  404179  
3   289068  310306  329562  353993  373325  391746  412644  
4   301524  322711  339925  364787  384123  402619  423494  
5        0       0       0       0       0       0       0  
..     ...     ...     ...     ...     ...     ...     ...  
84       0       0       0       0       0       0       0  
85    2721    2837    2927    3092    3182    3396    3421  
86    4316    4502    4642    4905    5082    5377    5497  
87    1850    1986    2093    2255    2388    2538    2621  
88    3618    3824    3862    4126    4316    4612    4758  

[88 rows x 11 columns]

<span style="color:red">Here, we have replaced the words after the regions name using regex pattern and assigned it to 'Region'. Also, strip() is used to get rid of the extra space in the begining of every columns.Similar process is carried out below to get rid of locations name using regex pattern and assigned it to 'Attribute'. After that the Attribute value is assigned to the 'Attribute' columns while Region value is assigned to 'Region' columns.</span>

In [808]:
#Removing the extra space infront of the columns name
Region=data['Year'].str.replace(r'(\s[CPF][a-z].+)','').str.strip()

In [809]:
Region

1            WHOLE COUNTRY
2            WHOLE COUNTRY
3            WHOLE COUNTRY
4            WHOLE COUNTRY
5     MA1 MAINLAND FINLAND
              ...         
84               MA2 ÅLAND
85              MK21 Åland
86              MK21 Åland
87              MK21 Åland
88              MK21 Åland
Name: Year, Length: 88, dtype: object

In [810]:
Attribute=data['Year'].str.replace(r'.+(?=(\s[CPF][a-z].+))','').str.strip()

In [811]:
Attribute

1       Citizens of foreign countries
2                 Persons born abroad
3           Foreign-language speakers
4     Persons with foreign background
5       Citizens of foreign countries
                   ...               
84    Persons with foreign background
85      Citizens of foreign countries
86                Persons born abroad
87          Foreign-language speakers
88    Persons with foreign background
Name: Year, Length: 88, dtype: object

In [812]:
data['Attribute']=Attribute

In [813]:
data['Region']=Region
data

0                                                Year    2010    2011    2012  \
1         WHOLE COUNTRY Citizens of foreign countries  167954  183133  195511   
2                   WHOLE COUNTRY Persons born abroad  248135  266148  285471   
3             WHOLE COUNTRY Foreign-language speakers  224388  244827  266949   
4       WHOLE COUNTRY Persons with foreign background  237066  257494  279616   
5   MA1 MAINLAND FINLAND Citizens of foreign count...       0       0       0   
..                                                ...     ...     ...     ...   
84          MA2 ÅLAND Persons with foreign background       0       0       0   
85           MK21 Åland Citizens of foreign countries    2415    2560    2610   
86                     MK21 Åland Persons born abroad    3785    4018    4171   
87               MK21 Åland Foreign-language speakers    1462    1623    1733   
88         MK21 Åland Persons with foreign background    3054    3293    3454   

0     2013    2014    2015    2016    2017    2018    2019  \
1   207511  219675  229765  243639  249452  257572  267629   
2   304279  321977  337162  357541  372802  387215  404179   
3   289068  310306  329562  353993  373325  391746  412644   
4   301524  322711  339925  364787  384123  402619  423494   
5        0       0       0       0       0       0       0   
..     ...     ...     ...     ...     ...     ...     ...   
84       0       0       0       0       0       0       0   
85    2721    2837    2927    3092    3182    3396    3421   
86    4316    4502    4642    4905    5082    5377    5497   
87    1850    1986    2093    2255    2388    2538    2621   
88    3618    3824    3862    4126    4316    4612    4758   

0                         Attribute                Region  
1     Citizens of foreign countries         WHOLE COUNTRY  
2               Persons born abroad         WHOLE COUNTRY  
3         Foreign-language speakers         WHOLE COUNTRY  
4   Persons with foreign background         WHOLE COUNTRY  
5     Citizens of foreign countries  MA1 MAINLAND FINLAND  
..                              ...                   ...  
84  Persons with foreign background             MA2 ÅLAND  
85    Citizens of foreign countries            MK21 Åland  
86              Persons born abroad            MK21 Åland  
87        Foreign-language speakers            MK21 Åland  
88  Persons with foreign background            MK21 Åland  

[88 rows x 13 columns]

As you can see from the above dataset we have unnecessary 'Year' column which carries the columns from original dataset as rows so it is better to remove 'Year' columns since we already have two new columns with 'Attribute' and 'Region'.

In [814]:
data.drop('Year',axis=1,inplace=True)

In [815]:
data

0     2010    2011    2012    2013    2014    2015    2016    2017    2018  \
1   167954  183133  195511  207511  219675  229765  243639  249452  257572   
2   248135  266148  285471  304279  321977  337162  357541  372802  387215   
3   224388  244827  266949  289068  310306  329562  353993  373325  391746   
4   237066  257494  279616  301524  322711  339925  364787  384123  402619   
5        0       0       0       0       0       0       0       0       0   
..     ...     ...     ...     ...     ...     ...     ...     ...     ...   
84       0       0       0       0       0       0       0       0       0   
85    2415    2560    2610    2721    2837    2927    3092    3182    3396   
86    3785    4018    4171    4316    4502    4642    4905    5082    5377   
87    1462    1623    1733    1850    1986    2093    2255    2388    2538   
88    3054    3293    3454    3618    3824    3862    4126    4316    4612   

0     2019                        Attribute                Region  
1   267629    Citizens of foreign countries         WHOLE COUNTRY  
2   404179              Persons born abroad         WHOLE COUNTRY  
3   412644        Foreign-language speakers         WHOLE COUNTRY  
4   423494  Persons with foreign background         WHOLE COUNTRY  
5        0    Citizens of foreign countries  MA1 MAINLAND FINLAND  
..     ...                              ...                   ...  
84       0  Persons with foreign background             MA2 ÅLAND  
85    3421    Citizens of foreign countries            MK21 Åland  
86    5497              Persons born abroad            MK21 Åland  
87    2621        Foreign-language speakers            MK21 Åland  
88    4758  Persons with foreign background            MK21 Åland  

[88 rows x 12 columns]

<b> Checking the nan values for the whole dataset

In [816]:
data.isna().sum().sum()

0

In [817]:
data2=data
data2

0     2010    2011    2012    2013    2014    2015    2016    2017    2018  \
1   167954  183133  195511  207511  219675  229765  243639  249452  257572   
2   248135  266148  285471  304279  321977  337162  357541  372802  387215   
3   224388  244827  266949  289068  310306  329562  353993  373325  391746   
4   237066  257494  279616  301524  322711  339925  364787  384123  402619   
5        0       0       0       0       0       0       0       0       0   
..     ...     ...     ...     ...     ...     ...     ...     ...     ...   
84       0       0       0       0       0       0       0       0       0   
85    2415    2560    2610    2721    2837    2927    3092    3182    3396   
86    3785    4018    4171    4316    4502    4642    4905    5082    5377   
87    1462    1623    1733    1850    1986    2093    2255    2388    2538   
88    3054    3293    3454    3618    3824    3862    4126    4316    4612   

0     2019                        Attribute                Region  
1   267629    Citizens of foreign countries         WHOLE COUNTRY  
2   404179              Persons born abroad         WHOLE COUNTRY  
3   412644        Foreign-language speakers         WHOLE COUNTRY  
4   423494  Persons with foreign background         WHOLE COUNTRY  
5        0    Citizens of foreign countries  MA1 MAINLAND FINLAND  
..     ...                              ...                   ...  
84       0  Persons with foreign background             MA2 ÅLAND  
85    3421    Citizens of foreign countries            MK21 Åland  
86    5497              Persons born abroad            MK21 Åland  
87    2621        Foreign-language speakers            MK21 Åland  
88    4758  Persons with foreign background            MK21 Åland  

[88 rows x 12 columns]

In [690]:
#grouping the dataset by Persons with foreign background attribute. Here, we can see the number of foreigners coming to Finland's different regions in different years.
a=data2[data2.Attribute=="Persons with foreign background"].groupby("Region").sum()
b=data2[data2.Attribute=="Citizens of foreign countries"].groupby("Region").sum()
a


0                          2010    2011    2012    2013    2014    2015  \
Region                                                                    
MA1 MAINLAND FINLAND          0       0       0       0       0       0   
MA2 ÅLAND                     0       0       0       0       0       0   
MK01 Uusimaa             126450  137712  150218  162693  175336  186621   
MK02 Southwest            21799   23239   25062   26658   28515   29883   
MK04 Satakunta             3964    4323    4873    5412    5853    6220   
MK05 Kanta-Häme            4050    4448    5042    5578    6019    6326   
MK06                      16114   17357   18465   19881   20936   21749   
MK07 Päijät-Häme           6933    7406    7964    8360    8741    9079   
MK08 Kymenlaakso           7368    8012    8766    9371    9855   10127   
MK09 South Karelia         5166    5618    6123    6567    6969    7320   
MK10 South Savo            3216    3413    3693    3933    4123    4349   
MK11 North Savo            4237    4711    5252    5672    6179    6479   
MK12 North Karelia         4184    4546    4884    5310    5627    5666   
MK13                       6116    6540    6981    7379    7786    8092   
MK14 South Ostrobothnia    2599    2962    3344    3724    3965    3995   
MK15 Ostrobothnia          8050    9005    9653   10342   11012   11591   
MK16                       1255    1382    1495    1627    1714    1886   
MK17 North Ostrobothnia    7016    7658    8191    8884    9440    9874   
MK18 Kainuu                1701    1767    1840    1905    1957    1956   
MK19 Lapland               3794    4102    4316    4610    4860    4850   
MK21 Åland                 3054    3293    3454    3618    3824    3862   
WHOLE COUNTRY            237066  257494  279616  301524  322711  339925   

0                          2016    2017    2018    2019  
Region                                                   
MA1 MAINLAND FINLAND          0       0       0       0  
MA2 ÅLAND                     0       0       0       0  
MK01 Uusimaa             200767  213290  225859  240435  
MK02 Southwest            31666   33506   35255   36828  
MK04 Satakunta             6904    7451    7803    8117  
MK05 Kanta-Häme            6758    7058    7115    7337  
MK06                      23249   24259   25226   26384  
MK07 Päijät-Häme           9863   10446   10931   11285  
MK08 Kymenlaakso          10770   10588   10532   10463  
MK09 South Karelia         7739    7977    8144    8351  
MK10 South Savo            4712    4737    4813    4945  
MK11 North Savo            6955    7234    7526    7664  
MK12 North Karelia         6042    6156    6188    6389  
MK13                       8841    9221    9412    9807  
MK14 South Ostrobothnia    4114    4193    4338    4591  
MK15 Ostrobothnia         12058   12681   13171   13614  
MK16                       2050    2058    2058    2123  
MK17 North Ostrobothnia   10943   11512   12054   12744  
MK18 Kainuu                2066    2109    2063    2154  
MK19 Lapland               5164    5331    5519    5505  
MK21 Åland                 4126    4316    4612    4758  
WHOLE COUNTRY            364787  384123  402619  423494

 <b> In above line of code, we are trying to find the Persons with foreign background  using groupby to group the data according to the regions. In the dataset, we can see that the number of Persons with foreign background moving to Finland has been increasing yearly.</b>
 
 <b> Same data is collected below using sort_values.

In [608]:

data2.sort_values(["Region","Attribute"]).groupby("Region").first()

0                          2010    2011    2012    2013    2014    2015  \
Region                                                                    
MA1 MAINLAND FINLAND          0       0       0       0       0       0   
MA2 ÅLAND                     0       0       0       0       0       0   
MK01 Uusimaa              86685   94966  102118  109213  116791  123300   
MK02 Southwest            14570   15553   16443   17084   18161   18834   
MK04 Satakunta             3016    3340    3805    4268    4631    4941   
MK05 Kanta-Häme            3045    3356    3836    4237    4566    4794   
MK06                      11411   12351   12663   13305   13738   14208   
MK07 Päijät-Häme           4924    5263    5595    5789    5894    5993   
MK08 Kymenlaakso           5494    5978    6411    6685    6949    6986   
MK09 South Karelia         3597    3908    4179    4421    4588    4764   
MK10 South Savo            2368    2479    2635    2752    2830    2997   
MK11 North Savo            3099    3470    3854    4128    4472    4697   
MK12 North Karelia         2987    3235    3400    3654    3747    3697   
MK13                       4349    4687    4902    5059    5303    5515   
MK14 South Ostrobothnia    2113    2434    2743    3004    3165    3147   
MK15 Ostrobothnia          6750    7608    7943    8252    8599    9002   
MK16                       1066    1193    1255    1293    1292    1424   
MK17 North Ostrobothnia    5558    6033    6286    6676    6983    7355   
MK18 Kainuu                1354    1360    1360    1328    1335    1357   
MK19 Lapland               3153    3359    3473    3642    3794    3827   
MK21 Åland                 2415    2560    2610    2721    2837    2927   
WHOLE COUNTRY            167954  183133  195511  207511  219675  229765   

0                          2016    2017    2018    2019  \
Region                                                    
MA1 MAINLAND FINLAND          0       0       0       0   
MA2 ÅLAND                     0       0       0       0   
MK01 Uusimaa             130963  135412  141597  149253   
MK02 Southwest            19814   20408   21185   21913   
MK04 Satakunta             5479    5833    6095    6280   
MK05 Kanta-Häme            5079    5206    5059    5118   
MK06                      15040   15308   15679   16329   
MK07 Päijät-Häme           6524    6730    7032    7134   
MK08 Kymenlaakso           7218    6706    6493    6270   
MK09 South Karelia         4974    4882    4776    4785   
MK10 South Savo            3241    3172    3191    3180   
MK11 North Savo            4974    5011    5091    5050   
MK12 North Karelia         3881    3819    3774    3909   
MK13                       6026    6152    6121    6309   
MK14 South Ostrobothnia    3180    3150    3227    3344   
MK15 Ostrobothnia          9174    9400    9519    9634   
MK16                       1537    1461    1425    1447   
MK17 North Ostrobothnia    8054    8231    8483    8908   
MK18 Kainuu                1395    1376    1327    1357   
MK19 Lapland               3994    4013    4102    3988   
MK21 Åland                 3092    3182    3396    3421   
WHOLE COUNTRY            243639  249452  257572  267629   

0                                            Attribute  
Region                                                  
MA1 MAINLAND FINLAND     Citizens of foreign countries  
MA2 ÅLAND                Citizens of foreign countries  
MK01 Uusimaa             Citizens of foreign countries  
MK02 Southwest           Citizens of foreign countries  
MK04 Satakunta           Citizens of foreign countries  
MK05 Kanta-Häme          Citizens of foreign countries  
MK06                     Citizens of foreign countries  
MK07 Päijät-Häme         Citizens of foreign countries  
MK08 Kymenlaakso         Citizens of foreign countries  
MK09 South Karelia       Citizens of foreign countries  
MK10 South Savo          Citizens of foreign countries  
MK11 North Savo          Citizens of fore

In [769]:
# In this line, we are trying to check which attributes have maximum value during the year 2010. For example, Uusimaa region have persons with foreign background as maximum values while whole country have persons born abroad as maximum values which is shown below. This can be useful to check if more foreigners moved to Finland or more people were born abroad and so on.
y=data2.loc[data2[2010]==data2.groupby("Region")[2010].transform("max")][["Region","Attribute",2010]].reset_index(drop=True)
y.loc[y.Attribute=='Persons with foreign background'][[2010,"Region"]].iplot(kind='bar',x="Region",title="Persons with foreign background ")

In [770]:
x=data2.loc[data2[2019]==data2.groupby("Region")[2019].transform("max")][["Region","Attribute",2019]].reset_index(drop=True)
x.loc[x.Attribute=='Persons with foreign background'][[2019,"Region"]].iplot(kind='bar',x="Region",title="Persons with foreign background ")


 <i> In this part, we tried to check the data for the year 2019 and 2010 and try to compare attribute for persons with foreign background and we can see that for the year 2010 only Uusimaa region have this attribute as maximum value which means rest of the regions have other attribute as maximum value while in year 2019 we can see along with Uusimaa region some other region like Southwest, Päijät-Häme, Kymenlaakso and others have this attribute as maximum values. This means that there have been increase in persons with foreign background throughout the years.

In [737]:
# Grouped using location to check every locations specific data.
grouped=data2.groupby(['Region','Attribute']).sum().reset_index()
grouped1=data2.groupby(['Attribute','Region']).sum()
grouped


0                 Region                        Attribute    2010    2011  \
0   MA1 MAINLAND FINLAND    Citizens of foreign countries       0       0   
1   MA1 MAINLAND FINLAND        Foreign-language speakers       0       0   
2   MA1 MAINLAND FINLAND              Persons born abroad       0       0   
3   MA1 MAINLAND FINLAND  Persons with foreign background       0       0   
4              MA2 ÅLAND    Citizens of foreign countries       0       0   
..                   ...                              ...     ...     ...   
83            MK21 Åland  Persons with foreign background    3054    3293   
84         WHOLE COUNTRY    Citizens of foreign countries  167954  183133   
85         WHOLE COUNTRY        Foreign-language speakers  224388  244827   
86         WHOLE COUNTRY              Persons born abroad  248135  266148   
87         WHOLE COUNTRY  Persons with foreign background  237066  257494   

0     2012    2013    2014    2015    2016    2017    2018    2019  
0        0       0       0       0       0       0       0       0  
1        0       0       0       0       0       0       0       0  
2        0       0       0       0       0       0       0       0  
3        0       0       0       0       0       0       0       0  
4        0       0       0       0       0       0       0       0  
..     ...     ...     ...     ...     ...     ...     ...     ...  
83    3454    3618    3824    3862    4126    4316    4612    4758  
84  195511  207511  219675  229765  243639  249452  257572  267629  
85  266949  289068  310306  329562  353993  373325  391746  412644  
86  285471  304279  321977  337162  357541  372802  387215  404179  
87  279616  301524  322711  339925  364787  384123  402619  423494  

[88 rows x 12 columns]

In [768]:
x1=grouped[grouped['Attribute']=='Foreign-language speakers'].reset_index().drop(21).drop(columns="index")
x1[[2010,2019,"Region"]].iplot(kind='line',x="Region",title="Foreign language speakers in 2010 and 2019")

 <i> We have grouped this data according to foreign-language speaker for the year 2010 and 2019 to see the change. Here, we can see that Uusimaa region's foreign language speakers data have more than doubled in the last 10 years and few other regions like Southwest, Kanta-Häme have also significant increment in the data. This suggests that along with Finnish language English language is also being accepted in those part of the country. Also all other regions also have increment in this value and foreign language speaker is only the data that have increased out of all the attribute throuthout the whole country i.e, why we choose to analyse this data.</i>

In [766]:
data2.Region.unique()

array(['WHOLE COUNTRY', 'MA1 MAINLAND FINLAND', 'MK01 Uusimaa',
       'MK02 Southwest', 'MK04 Satakunta', 'MK05 Kanta-Häme', 'MK06',
       'MK07 Päijät-Häme', 'MK08 Kymenlaakso', 'MK09 South Karelia',
       'MK10 South Savo', 'MK11 North Savo', 'MK12 North Karelia', 'MK13',
       'MK14 South Ostrobothnia', 'MK15 Ostrobothnia', 'MK16',
       'MK17 North Ostrobothnia', 'MK18 Kainuu', 'MK19 Lapland',
       'MA2 ÅLAND', 'MK21 Åland'], dtype=object)

Using loc[ ] to see the citizens of foreign countries for Uusimaa region for all the years.

In [776]:
# Plotting the graph to see the trend of foreigners coming to Uusimaa region.
grouped1.loc["Citizens of foreign countries",2019].iplot(title='Citizens of foreign countries')

<b> This graph shows the citizens of foreign countries for the year 2019 throughout every region in Finland.

In [778]:
grouped1.loc["Citizens of foreign countries","MK01 Uusimaa"].iplot(title='Citizens of foreign countries')

 <b> Here, we have used matplotlib for visualization purposes. Above, we are trying to show the citizens of foreign countries coming to Uusimaa region only from year 2010 to 2019. The data shows that this has almost doubled over the course of 10 years period. 

In [781]:
# Defining a function 
def plot_region(att,reg):
    grouped1.loc[att,reg].iplot()

In [843]:
# Plotting the graph
interact(plot_region,att=data2.Attribute.unique(),reg=data2.Region.unique(),)
plt.show()

interactive(children=(Dropdown(description='att', options=('Citizens of foreign countries', 'Persons born abro…

<b><span style='color:red'> Here, the graph is presented with attribute and region box where the user can select the desired location and attribute. X-axis represents the years while y-axis is numerical value. When selected the whole country as region and citizens of foreign countries we can see in 2019 the number was 267,629 while in 2018 the number was 257,572. In this way, comparisons can be made during different years for different regions. It's more convienent to look at the graph and see the trend than looking at the dataset which makes it confusing to figure out the trend.</span></b> 

In [789]:
data2

0     2010    2011    2012    2013    2014    2015    2016    2017    2018  \
1   167954  183133  195511  207511  219675  229765  243639  249452  257572   
2   248135  266148  285471  304279  321977  337162  357541  372802  387215   
3   224388  244827  266949  289068  310306  329562  353993  373325  391746   
4   237066  257494  279616  301524  322711  339925  364787  384123  402619   
5        0       0       0       0       0       0       0       0       0   
..     ...     ...     ...     ...     ...     ...     ...     ...     ...   
84       0       0       0       0       0       0       0       0       0   
85    2415    2560    2610    2721    2837    2927    3092    3182    3396   
86    3785    4018    4171    4316    4502    4642    4905    5082    5377   
87    1462    1623    1733    1850    1986    2093    2255    2388    2538   
88    3054    3293    3454    3618    3824    3862    4126    4316    4612   

0     2019                        Attribute                Region  
1   267629    Citizens of foreign countries         WHOLE COUNTRY  
2   404179              Persons born abroad         WHOLE COUNTRY  
3   412644        Foreign-language speakers         WHOLE COUNTRY  
4   423494  Persons with foreign background         WHOLE COUNTRY  
5        0    Citizens of foreign countries  MA1 MAINLAND FINLAND  
..     ...                              ...                   ...  
84       0  Persons with foreign background             MA2 ÅLAND  
85    3421    Citizens of foreign countries            MK21 Åland  
86    5497              Persons born abroad            MK21 Åland  
87    2621        Foreign-language speakers            MK21 Åland  
88    4758  Persons with foreign background            MK21 Åland  

[88 rows x 12 columns]

In [791]:
whole_country=data2.query('Region=="WHOLE COUNTRY"').reset_index(drop=True).iplot(kind='bar',x='Attribute',y=[2010,2019],title='whole country data')
whole_country

<b> This visualization shows the overall change in all the attributes from year 2010 to 2019 for the whole country. The increase in all the attribute suggests that over the course of 10 years Finland have been more leaning towards internationalization of its society. Perhaps, this suggests that Finland have been one of the few countries in Europe to mix its culture with international culture to bring variety.

In [792]:
region

0    2010   2011    2012    2013    2014    2015    2016    2017    2018  \
0       0      0       0       0       0       0       0       0       0   
1       0      0       0       0       0       0       0       0       0   
2       0      0       0       0       0       0       0       0       0   
3       0      0       0       0       0       0       0       0       0   
4   86685  94966  102118  109213  116791  123300  130963  135412  141597   
..    ...    ...     ...     ...     ...     ...     ...     ...     ...   
79      0      0       0       0       0       0       0       0       0   
80   2415   2560    2610    2721    2837    2927    3092    3182    3396   
81   3785   4018    4171    4316    4502    4642    4905    5082    5377   
82   1462   1623    1733    1850    1986    2093    2255    2388    2538   
83   3054   3293    3454    3618    3824    3862    4126    4316    4612   

0     2019                        Attribute                Region  
0        0    Citizens of foreign countries  MA1 MAINLAND FINLAND  
1        0              Persons born abroad  MA1 MAINLAND FINLAND  
2        0        Foreign-language speakers  MA1 MAINLAND FINLAND  
3        0  Persons with foreign background  MA1 MAINLAND FINLAND  
4   149253    Citizens of foreign countries          MK01 Uusimaa  
..     ...                              ...                   ...  
79       0  Persons with foreign background             MA2 ÅLAND  
80    3421    Citizens of foreign countries            MK21 Åland  
81    5497              Persons born abroad            MK21 Åland  
82    2621        Foreign-language speakers            MK21 Åland  
83    4758  Persons with foreign background            MK21 Åland  

[84 rows x 12 columns]

<b><span style='color:blue'> This is the bar graph visualization of different attributes for the whole country from 2010 and 2019. Specific year can be selected to see the attributes contribution or all the years data can be viewed together but in doing so the graphical visualization becomes a bit messier so it is recomended to select specific years from the index to view the graph.</span>

In [873]:
region=data2.query('Region!="WHOLE COUNTRY"').reset_index(drop=True).iplot(kind='line',x=["Attribute","Region"],y=[2019])



<b> As this is the graphical presentation for all the regions of Finland it is not possible to get proper view and understanding of the data. This is because there are more than 80 rows which is difficult to fit in the graph axis. 

# Summary
Through this data analysis we can summarize that almost all the attributes have doubled in the last 10 years. This also indicates that Finland has been more involved in bringing international cultures. There might be other factors affecting the increment of foreign population like international students coming to study, importing more work labour force and so on. 